### PR TITLE
PR: Use PIPE for stdout and stderr (IPython console) and improve comm connection

### PR DESCRIPTION
**Pull request by impact27**
## Description of Changes

- Removes the std files that Spyder uses for stdout, stderr and faulthandler. They are replaced by PIPEs and therefore do not require cleaning up or complex file I/O code.
- Goes with spyder-ide/spyder-kernels#412

Fixes https://github.com/spyder-ide/spyder/issues/13336

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct:

### CODE REVIEW

**Comment by ccordoba12** [comment]
@impact27, is this ready for review?

**Comment by impact27** [comment]
@ccordoba12 this is ready for review

File: spyder/plugins/ipythonconsole/widgets/client.py [hunk 1]
```diff
@@ -363,74 +356,49 @@ def connection_file(self):
...
+            return
+        if self.shellwidget.isHidden():
+            error_text = '<tt>%s</tt>' % stderr
+            # Avoid printing the same thing again
+            if self.error_text != error_text:
+                if self.error_text:
+                    # Append to error text
+                    error_text = self.error_text + error_text
+                self.show_kernel_error(error_text)
+        if self.shellwidget._starting:
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/client.py:384]:
```suggestion

        if self.shellwidget._starting:
```

**Review by ccordoba12 — COMMENTED** [review]
Thanks @impact27 for your work on this! It helps to simplify our code a lot!

**Comment by impact27** [comment]
@ccordoba12 small change: ad32bfa uses a Qt Signal instead of a callback for faults (because this is the Qt way and a callback was kind of ugly)

**Comment by impact27** [comment]
@ccordoba12 the comm have been changed on this side as well to match https://github.com/spyder-ide/spyder-kernels/pull/412

**Comment by ccordoba12** [comment]
@impact27, this has now a small merge conflict due to PR #19646. To fix it please

* Remove the `TODO` I left for you:

    https://github.com/spyder-ide/spyder/blob/ed59291b998ff97889cf0fecab27a8d252892e59/spyder/plugins/ipythonconsole/utils/kernel_handler.py#L151-L159

* Make sure that `test_show_spyder_kernels_error_on_restart` passes after fixing that TODO and remove the `xfail` decorator from it.

    https://github.com/spyder-ide/spyder/blob/ed59291b998ff97889cf0fecab27a8d252892e59/spyder/plugins/ipythonconsole/tests/test_ipythonconsole.py#L2504-L2508

**Comment by impact27** [comment]
@ccordoba12 this is ready for review

File: spyder/plugins/ipythonconsole/widgets/shell.py [hunk 2]
```diff
@@ -295,44 +282,61 @@ def is_external_kernel(self):
...
 
-    def pop_execute_queue(self):
-        """Pop one waiting instruction."""
-        if self._execute_queue:
-            self.execute(*self._execute_queue.pop(0))
-
     def send_spyder_kernel_configuration(self):
         """Send kernel configuration to spyder kernel."""
+        # Set current cwd
+        self.set_cwd()
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/shell.py:318]: Why is this necessary here?

**Review by ccordoba12 — COMMENTED** [review]
Sorry for the delay in reviewing this one @impact27, I got side-tracked by PR #19937.

Besides my usual style suggestions, my concern is about `sig_kernel_connection_state`, which I found a bit confusing (as explained below).

File: spyder/plugins/ipythonconsole/widgets/client.py [hunk 3]
```diff
@@ -172,39 +174,25 @@ def __del__(self):
             self.restart_thread.wait()
 
     # ----- Private methods ---------------------------------------------------
-    def _before_prompt_is_ready(self):
```
[the remark this answers is not in the document]
Comment (impact27) [reply spyder/plugins/ipythonconsole/widgets/client.py:175]: because sig_prompt_ready and sig_remote_execute are replaced, so the only thing that this function does is call _show_loading_page, so call to this function can be replaced by _show_loading_page

File: spyder/plugins/ipythonconsole/widgets/shell.py (hunk 2 above)
Comment (impact27) [reply spyder/plugins/ipythonconsole/widgets/shell.py:327]: If the kernel restarts, the kernel loses the cwd, so it must be reset here.

**Review by ccordoba12 — COMMENTED** [review]
Last minor suggestions, then this should be ready @impact27.

**Comment by ccordoba12** [comment]
@impact27, please re-sync our Spyder-kernels subrepo. That will make the conda-based installers workflows to pass and then I'll merge.

**Review by ccordoba12 — APPROVED** [review]
Thanks @impact27 for your hard work on this one!
